src/main/java/engineering/swat/watch/impl/JDKRecursiveDirectoryWatcher.java

```diff
@@ -0,0 +1,317 @@
+            });
+    }
+
+    private void handleDeleteDirectory(WatchEvent ev) {
+        var removedPath = ev.calculateFullPath();
+        try {
+            var existingWatch = activeWatches.remove(removedPath);
+            if (existingWatch != null) {
+                logger.debug("Clearing watch on removed directory: {}", removedPath);
+                existingWatch.close();
+            }
+        } catch (IOException ex) {
+            logger.error("Error clearing: {} {}", removedPath, ex);
+        }
+    }
+
+    /** Only register a watched for every sub directory */
+    private class InitialDirectoryScan extends SimpleFileVisitor<Path> {
+        protected final Path subRoot;
+
+        public InitialDirectoryScan(Path root) {
+            this.subRoot = root;
+        }
+        @Override
+        public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
+            logger.error("We could not visit {} to schedule recursive file watches: {}", file, exc);
+            return FileVisitResult.CONTINUE;
+        }
+
+        @Override
+        public FileVisitResult preVisitDirectory(Path subdir, BasicFileAttributes attrs) throws IOException {
+            addNewDirectory(subdir);
+            return FileVisitResult.CONTINUE;
+        }
+
+        @Override
+        public FileVisitResult postVisitDirectory(Path subdir, IOException exc) throws IOException {
+            if (exc != null) {
+                logger.error("Error during directory iteration: {} = {}", subdir, exc);
+            }
+            return FileVisitResult.CONTINUE;
+        }
+
+        private void addNewDirectory(Path dir) throws IOException {
+            var watcher = activeWatches.computeIfAbsent(dir, d -> new JDKDirectoryWatcher(d, exec, relocater(dir)));
+            try {
+                if (!watcher.safeStart()) {
+                    logger.debug("We lost the race on starting a nested watcher, that shouldn't be a problem, but its a very busy, so we might have lost a few events in {}", dir);
+                }
+            } catch (IOException ex) {
+                activeWatches.remove(dir);
+                logger.error("Could not register a watch for: {} ({})", dir, ex);
+                throw ex;
+            }
+        }
+
+        /** Make sure that the events are relative to the actual root of the recursive watcher */
+        private Consumer<WatchEvent> relocater(Path subRoot) {
+            final Path newRelative = root.relativize(subRoot);
+            return ev -> {
+                var rewritten = new WatchEvent(ev.getKind(), root, newRelative.resolve(ev.getRelativePath()));
+                processEvents(rewritten);
+            };
+        }
+    }
+
+    /** register watch for new sub-dir, but also simulate event for every file & subdir found */
+    private class NewDirectoryScan extends InitialDirectoryScan {
+        protected final List<WatchEvent> events;
+        protected final Set<Path> seenFiles;
```

Comment:
`seenFiles` is never updated (and initially empty)

Reply:
good spotted, that's a bug.